File: EVANS/frmMantCliente.vb
Imports System.Data.SqlClient

Public Class frmMantCliente

    Dim objCliente As New clsCliente
    Dim bolAccion As Boolean = True

    Private Sub frmMantCliente_FormClosing(ByVal sender As Object, ByVal e As System.Windows.Forms.FormClosingEventArgs) Handles Me.FormClosing

        Try
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If

            LimpiarRAM()

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub frmMantCliente_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        Try
            Me.Top = 0
            Me.Left = 0

            TabControl1.TabPages(1).Enabled = False
            txtBuscar.Enabled = False

            btnGrabar.Enabled = False
            btnCancelar.Enabled = False
            btnEditar.Enabled = False
            btnBuscar.Enabled = False

            optBuscar.Checked = True

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End Try

    End Sub

    Private Sub optTodos_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles optTodos.CheckedChanged

        Try
            txtBuscar.Clear()
            dgvListado.Rows.Clear()
            Dim lstLista As New List(Of clsCliente)

            lstLista = objCliente.Listar()

            For i As Integer = 0 To lstLista.Count - 1
                dgvListado.Rows.Add()
                dgvListado.Rows(i).Cells(0).Value = lstLista(i).Codigo
                dgvListado.Rows(i).Cells(1).Value = lstLista(i).Nombre
                dgvListado.Rows(i).Cells(2).Value = lstLista(i).TipoID.Descripcion
                dgvListado.Rows(i).Cells(3).Value = lstLista(i).NumeroID
            Next

            txtBuscar.Enabled = False
            btnBuscar.Enabled = False

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub optBuscar_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles optBuscar.CheckedChanged, optNro.CheckedChanged

        Try
            dgvListado.Rows.Clear()
            txtBuscar.Enabled = True
            txtBuscar.Focus()
            btnBuscar.Enabled = True

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub btnNuevo_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnNuevo.Click

        Try
            objCliente = New clsCliente

            'Comportamiento del control TabControl
            TabControl1.TabPages(1).Enabled = True
            TabControl1.TabPages(0).Enabled = False
            TabControl1.SelectTab(1)
            txtNombre.Focus()

            'Limpiar y activa cajas de texto
            LimpiarControles(TabControl1, 1)
            DesactivarControles(TabControl1, 1, False)
            txtCodigo.ReadOnly = True

            'Comportamiento de botones
            btnNuevo.Enabled = False
            btnGrabar.Enabled = True
            btnCancelar.Enabled = True

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub btnGrabar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnGrabar.Click

        Try
            If txtNombre.Text = "" Or cbTipoID.Text = "" Or txtNroID.Text = "" Then
                MessageBox.Show("Datos incompletos", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
                Return
            End If

            objCliente.Nombre = UCase(txtNombre.Text)
            objCliente.TipoID.Codigo = BuscarCodigo("iden_codigo", "iden_descripcion", cbTipoID.Text, "TipoIdentificacion")
            objCliente.NumeroID = txtNroID.Text
            objCliente.Telefono = txtTelefono.Text
            objCliente.Fax = txtFax.Text
            objCliente.Email = txtEmail.Text
            objCliente.Representante = txtRepresentante.Text

            objCliente.Direccion.Clear()
            Dim i As Integer
            For i = 0 To dgvDireccion.Rows.Count - 2
                Dim item As New clsDireccion
                item.Direccion = IIf(dgvDireccion.Rows(i).Cells(0).Value = Nothing, "", dgvDireccion.Rows(i).Cells(0).Value)
                item.Ciudad = IIf(dgvDireccion.Rows(i).Cells(1).Value = Nothing, "", dgvDireccion.Rows(i).Cells(1).Value)
                item.Provincia = IIf(dgvDireccion.Rows(i).Cells(2).Value = Nothing, "", dgvDireccion.Rows(i).Cells(2).Value)
                objCliente.Direccion.Add(item)
                item = Nothing
            Next

            If bolAccion = True Then
                If objCliente.Grabar() = False Then
                    Return
                End If
            Else
                If objCliente.Actualizar() = False Then
                    Return
                Else
                    bolAccion = True
                End If
            End If

            'Comportamiento del TabControl
            TabControl1.TabPages(0).Enabled = True
            TabControl1.TabPages(1).Enabled = False
            TabControl1.SelectTab(0)

            'Comportamiento de Botones
            btnNuevo.Enabled = True
            btnGrabar.Enabled = False
            btnEditar.Enabled = False
            btnCancelar.Enabled = False

            btnGrabar.Text = "Grabar"

            optBuscar.Checked = True

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub btnEditar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnEditar.Click

        Try
            'Activa las cajas de texto
            DesactivarControles(TabControl1, 1, False)
            txtCodigo.ReadOnly = True
            txtNombre.Focus()

            'Comportamiento de botones
            btnNuevo.Enabled = False
            btnGrabar.Enabled = True
            btnEditar.Enabled = False
            btnCancelar.Enabled = True

            btnGrabar.Text = "Actualizar"
            bolAccion = False

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub btnCancelar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnCancelar.Click

        Try
            'Comportamiento del TabControl
            TabControl1.TabPages(0).Enabled = True
            TabControl1.TabPages(1).Enabled = False
            TabControl1.SelectTab(0)

            bolAccion = True
            btnGrabar.Text = "Grabar"

            'Comportamiento de botones
            btnNuevo.Enabled = True
            btnGrabar.Enabled = False
            btnEditar.Enabled = False
            btnCancelar.Enabled = False

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub btnSalir_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSalir.Click
        Try
            Me.Close()

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub btnBuscar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnBuscar.Click

        Try
            If txtBuscar.Text = "" Then
                dgvListado.Rows.Clear()
                MessageBox.Show("Ingrese nombre a buscar", "Datos insuficientes", MessageBoxButtons.OK, MessageBoxIcon.Information)
                txtBuscar.Focus()
            Else

                If optBuscar.Checked = True Then
                    objCliente.Nombre = txtBuscar.Text
                    Dim drLista As SqlDataReader
                    drLista = objCliente.BuscarXNombre()
                    dgvListado.Rows.Clear()
                    Dim i As Integer = 0
                    While drLista.Read()
                        dgvListado.Rows.Add()
                        dgvListado.Rows(i).Cells(0).Value = NullToString(drLista(0))
                        dgvListado.Rows(i).Cells(1).Value = NullToString(drLista(1))
                        dgvListado.Rows(i).Cells(2).Value = NullToString(drLista(2))
                        dgvListado.Rows(i).Cells(3).Value = NullToString(drLista(3))
                        i = i + 1
                    End While
                    drLista.Close()
                End If

                If optNro.Checked = True Then
                    dgvListado.Rows.Clear()
                    objCliente = objCliente.BuscarXID(txtBuscar.Text)
                    dgvListado.Rows.Add()
                    dgvListado.Rows(0).Cells(0).Value = objCliente.Codigo
                    dgvListado.Rows(0).Cells(1).Value = objCliente.Nombre
                    dgvListado.Rows(0).Cells(2).Value = objCliente.TipoID.Descripcion
                    dgvListado.Rows(0).Cells(3).Value = objCliente.NumeroID
                End If

            End If

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub txtBuscar_KeyPress(ByVal sender As Object, ByVal e As System.Windows.Forms.KeyPressEventArgs) Handles txtBuscar.KeyPress

        Try
            If e.KeyChar = ChrW(Keys.Enter) Then
                e.Handled = True
                btnBuscar.PerformClick()
            End If

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub cbTipoID_DropDown(ByVal sender As Object, ByVal e As System.EventArgs) Handles cbTipoID.DropDown

        Try
            LLenarCombo("iden_codigo", "iden_descripcion", "TipoIdentificacion", cbTipoID, False)

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub cbTipoID_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cbTipoID.SelectedIndexChanged

        Try
            txtNroID.Clear()

            If LCase(cbTipoID.Text) = "sin documento" Then
                txtNroID.Text = "00000000"
            End If

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub txtNroID_GotFocus(ByVal sender As Object, ByVal e As System.EventArgs) Handles txtNroID.GotFocus

        Try
            txtNroID.MaxLength = 0

            If LCase(cbTipoID.Text) = "dni" Then
                txtNroID.MaxLength = 8
            End If

            If LCase(cbTipoID.Text) = "ruc" Then
                txtNroID.MaxLength = 11
            End If

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub dgvListado_DoubleClick(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles dgvListado.DoubleClick

        Try
            Dim strCodigo As String
            strCodigo = dgvListado.SelectedRows(0).Cells(0).Value

            objCliente.Codigo = strCodigo
            If objCliente.BuscarXCodigo() Then

                TabControl1.TabPages(1).Enabled = True
                TabControl1.TabPages(0).Enabled = False
                TabControl1.SelectTab(1)

                txtCodigo.Text = objCliente.Codigo
                txtNombre.Text = objCliente.Nombre
                cbTipoID.Text = objCliente.TipoID.Descripcion
                txtNroID.Text = objCliente.NumeroID
                txtTelefono.Text = objCliente.Telefono
                dgvDireccion.Rows.Clear()

                For i As Integer = 0 To objCliente.Direccion.Count - 1
                    dgvDireccion.Rows.Add()
                    dgvDireccion.Rows(i).Cells(0).Value = IIf(objCliente.Direccion(i).Direccion = "", Nothing, objCliente.Direccion(i).Direccion)
                    dgvDireccion.Rows(i).Cells(1).Value = IIf(objCliente.Direccion(i).Ciudad = "", Nothing, objCliente.Direccion(i).Ciudad)
                    dgvDireccion.Rows(i).Cells(2).Value = IIf(objCliente.Direccion(i).Provincia = "", Nothing, objCliente.Direccion(i).Provincia)
                Next

                'Desactivar controles
                DesactivarControles(TabControl1, 1, True)

                'Comportamiento de botones
                btnNuevo.Enabled = True
                btnGrabar.Enabled = False
                btnEditar.Enabled = True
                btnCancelar.Enabled = True

                btnGrabar.Text = "Grabar"

            Else
                Return
            End If

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try


    End Sub

    Private Sub txtBuscarCodigo_KeyPress(ByVal sender As System.Object, ByVal e As System.Windows.Forms.KeyPressEventArgs) Handles txtBuscarCodigo.KeyPress

        Try
            If e.KeyChar = ChrW(Keys.Enter) Then
                e.Handled = True
                objCliente = New clsCliente
                dgvListado.Rows.Clear()
                objCliente.Codigo = txtBuscarCodigo.Text
                If objCliente.BuscarXCodigo = True Then
                    If objCliente.Nombre <> "" Then
                        dgvListado.Rows.Add()
                        dgvListado.Rows(0).Cells(0).Value = objCliente.Codigo
                        dgvListado.Rows(0).Cells(1).Value = objCliente.Nombre
                        dgvListado.Rows(0).Cells(2).Value = objCliente.TipoID.Descripcion
                        dgvListado.Rows(0).Cells(3).Value = objCliente.NumeroID
                    End If
                Else
                    dgvListado.Rows.Clear()
                End If
            End If

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
        End Try

    End Sub

    Private Sub TabControl1_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles TabControl1.SelectedIndexChanged

        If TabControl1.SelectedIndex = 1 Then
            txtBuscarCodigo.Enabled = False
        Else
            txtBuscarCodigo.Enabled = True
        End If
    End Sub
End Class

File: src/EVANS.UI.WinForms/Catalogo/frmMantCliente.cs
using System.Globalization;
using EVANS.Application.Catalogo.Commands;
using EVANS.Application.Catalogo.DTOs;
using EVANS.Application.Catalogo.Queries;
using MediatR;

namespace EVANS.UI.WinForms.Catalogo;

public partial class frmMantCliente : Form
{
    private readonly IMediator? _mediator;
    private IReadOnlyList<ClienteDto> _clientes = [];
    private IReadOnlyList<TipoIdentificacionDto> _tiposIdentificacion = [];
    private bool _isCreateMode = true;

    public frmMantCliente() => InitializeComponent();

    public frmMantCliente(IMediator mediator)
    {
        _mediator = mediator;
        InitializeComponent();
        WireEvents();
        ApplyInitialState();
    }

    internal int SelectedTabIndex => tabControl1.SelectedIndex;
    internal bool DetailsTabEnabled => tabPageDetalles.Enabled;
    internal bool ListingTabEnabled => tabPageListado.Enabled;
    internal bool BuscarEnabled => btnBuscar.Enabled;
    internal bool BuscarTextEnabled => txtBuscar.Enabled;
    internal bool NuevoEnabled => btnNuevo.Enabled;
    internal bool GrabarEnabled => btnGrabar.Enabled;
    internal bool EditarEnabled => btnEditar.Enabled;
    internal bool CancelarEnabled => btnCancelar.Enabled;
    internal string GrabarText => btnGrabar.Text;
    internal string CodigoText => txtCodigo.Text;
    internal string StatusMessage => lblmsg.Text;
    internal int ListCount => dgvListado.Rows.Cast<DataGridViewRow>().Count(row => !row.IsNewRow);
    internal string? FirstListNombre => ListCount == 0 ? null : Convert.ToString(dgvListado.Rows[0].Cells[1].Value, CultureInfo.CurrentCulture);
    internal string? FirstListNumero => ListCount == 0 ? null : Convert.ToString(dgvListado.Rows[0].Cells[3].Value, CultureInfo.CurrentCulture);
    internal void SetTestSearchText(string value) => txtBuscar.Text = value;
    internal void SetTestNombre(string value) => txtNombre.Text = value;
    internal void SetTestNroId(string value) => txtNroID.Text = value;
    internal void SetTestTelefono(string value) => txtTelefono.Text = value;
    internal void SetTestFax(string value) => txtFax.Text = value;
    internal void SetTestEmail(string value) => txtEmail.Text = value;
    internal void SetTestRepresentante(string value) => txtRepresentante.Text = value;
    internal void SetTestTipoId(int codigo) => SelectTipoIdentificacion(codigo);
    internal void AddTestDireccion(string calle, string ciudad, string provincia) => dgvDireccion.Rows.Add(calle, ciudad, provincia);
    internal void SelectAllSearchForTest() => optTodos.Checked = true;
    internal void SelectNameSearchForTest() => optBuscar.Checked = true;
    internal void SelectNumberSearchForTest() => optNro.Checked = true;
    internal void BeginNewForTest() => btnNuevo_Click();
    internal void BeginEditForTest() => btnEditar_Click();

    internal async Task OpenFirstForTestAsync()
    {
        if (ListCount > 0)
            await OpenClienteAsync(Convert.ToInt32(dgvListado.Rows[0].Cells[0].Value, CultureInfo.InvariantCulture));
    }

    internal async Task LoadClientesAsync()
    {
        await EnsureTiposIdentificacionAsync();
        _clientes = await Mediator.Send(new ListClientesQuery());
        BindClientes(_clientes);
    }

    internal async Task<bool> BuscarAsync(bool showMessages = false)
    {
        if (string.IsNullOrWhiteSpace(txtBuscar.Text))
        {
            dgvListado.Rows.Clear();
            SetStatus("Ingrese nombre a buscar");
            if (showMessages)
                MessageBox.Show("Ingrese nombre a buscar", "Datos insuficientes", MessageBoxButtons.OK, MessageBoxIcon.Information);
            txtBuscar.Focus();
            return false;
        }

        if (_clientes.Count == 0)
            _clientes = await Mediator.Send(new ListClientesQuery());

        var search = txtBuscar.Text.Trim();
        var filtered = optNro.Checked
            ? _clientes.Where(cliente => string.Equals(cliente.NroIdentificacion, search, StringComparison.CurrentCultureIgnoreCase))
            : _clientes.Where(cliente => cliente.RazonSocial.StartsWith(search, StringComparison.CurrentCultureIgnoreCase));

        BindClientes(filtered);
        SetStatus("Los campos marcados con asterisco (*) son obligatorios");
        return true;
    }

    internal async Task<bool> SaveAsync(bool showMessages = false)
    {
        var tipoIdentificacion = ResolveTipoIdentificacion();
        if (string.IsNullOrWhiteSpace(txtNombre.Text) || tipoIdentificacion is null || string.IsNullOrWhiteSpace(txtNroID.Text))
        {
            SetStatus("Datos incompletos");
            if (showMessages)
                MessageBox.Show("Datos incompletos", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            return false;
        }

        var razonSocial = txtNombre.Text.Trim().ToUpperInvariant();
        var numeroIdentificacion = txtNroID.Text.Trim();
        var direcciones = ReadDirecciones();

        if (_isCreateMode)
        {
            var result = await Mediator.Send(new CreateClienteCommand(
                razonSocial,
                tipoIdentificacion.Codigo,
                numeroIdentificacion,
                tipoIdentificacion.LongitudRequerida,
                EmptyToNull(txtTelefono.Text),
                EmptyToNull(txtFax.Text),
                EmptyToNull(txtEmail.Text),
                EmptyToNull(txtRepresentante.Text),
                direcciones));
            if (!result.IsSuccess) return ShowFailure(result.Error, showMessages);
        }
        else
        {
            var result = await Mediator.Send(new UpdateClienteCommand(
                int.Parse(txtCodigo.Text, CultureInfo.InvariantCulture),
                razonSocial,
                tipoIdentificacion.Codigo,
                numeroIdentificacion,
                tipoIdentificacion.LongitudRequerida,
                EmptyToNull(txtTelefono.Text),
                EmptyToNull(txtFax.Text),
                EmptyToNull(txtEmail.Text),
                EmptyToNull(txtRepresentante.Text),
                direcciones));
            if (!result.IsSuccess) return ShowFailure(result.Error, showMessages);
            _isCreateMode = true;
        }

        await ReturnToListingAsync();
        return true;
    }

    private void WireEvents()
    {
        Load += async (_, _) => await EnsureTiposIdentificacionAsync();
        optTodos.CheckedChanged += async (_, _) => await optTodos_CheckedChanged();
        optBuscar.CheckedChanged += (_, _) => SearchOption_CheckedChanged(optBuscar);
        optNro.CheckedChanged += (_, _) => SearchOption_CheckedChanged(optNro);
        dgvListado.DoubleClick += async (_, _) => await dgvListado_DoubleClick();
        btnNuevo.Click += (_, _) => btnNuevo_Click();
        btnGrabar.Click += async (_, _) => await SaveAsync(showMessages: true);
        btnEditar.Click += (_, _) => btnEditar_Click();
        btnCancelar.Click += async (_, _) => await ReturnToListingAsync();
        btnSalir.Click += (_, _) => Close();
        btnBuscar.Click += async (_, _) => await BuscarAsync(showMessages: true);
        txtBuscar.KeyPress += async (_, e) =>
        {
            if (e.KeyChar == (char)Keys.Enter)
            {
                e.Handled = true;
                await BuscarAsync(showMessages: true);
            }
        };
        cbTipoID.DropDown += async (_, _) => await EnsureTiposIdentificacionAsync();
        cbTipoID.SelectedIndexChanged += (_, _) => cbTipoID_SelectedIndexChanged();
        txtNroID.GotFocus += (_, _) => txtNroID_GotFocus();
        txtBuscarCodigo.KeyPress += async (_, e) => await txtBuscarCodigo_KeyPress(e);
        tabControl1.SelectedIndexChanged += (_, _) => txtBuscarCodigo.Enabled = tabControl1.SelectedIndex == 0;
    }

    private void ApplyInitialState()
    {
        Top = 0;
        Left = 0;
        tabPageDetalles.Enabled = false;
        txtBuscar.Enabled = false;
        btnGrabar.Enabled = false;
        btnCancelar.Enabled = false;
        btnEditar.Enabled = false;
        btnBuscar.Enabled = false;
        optBuscar.Checked = true;
    }

    private async Task optTodos_CheckedChanged()
    {
        if (!optTodos.Checked) return;
        txtBuscar.Clear();
        await EnsureTiposIdentificacionAsync();
        if (_clientes.Count == 0)
            _clientes = await Mediator.Send(new ListClientesQuery());
        BindClientes(_clientes);
        txtBuscar.Enabled = false;
        btnBuscar.Enabled = false;
    }

    private void SearchOption_CheckedChanged(RadioButton option)
    {
        if (!option.Checked) return;
        dgvListado.Rows.Clear();
        txtBuscar.Enabled = true;
        txtBuscar.Focus();
        btnBuscar.Enabled = true;
    }

    private async Task dgvListado_DoubleClick()
    {
        var row = dgvListado.SelectedRows.Cast<DataGridViewRow>().FirstOrDefault(row => !row.IsNewRow)
            ?? dgvListado.CurrentRow;
        if (row is not null && !row.IsNewRow)
            await OpenClienteAsync(Convert.ToInt32(row.Cells[0].Value, CultureInfo.InvariantCulture));
    }

    private async Task OpenClienteAsync(int codigo)
    {
        await EnsureTiposIdentificacionAsync();
        var cliente = await Mediator.Send(new GetClienteByIdQuery(codigo));
        if (cliente is null) return;

        tabPageDetalles.Enabled = true;
        tabPageListado.Enabled = false;
        tabControl1.SelectTab(tabPageDetalles);
        txtCodigo.Text = cliente.Codigo.ToString(CultureInfo.InvariantCulture);
        txtNombre.Text = cliente.RazonSocial;
        SelectTipoIdentificacion(cliente.TipoIdCodigo);
        txtNroID.Text = cliente.NroIdentificacion;
        txtTelefono.Text = cliente.Telefono ?? string.Empty;
        txtFax.Text = cliente.Fax ?? string.Empty;
        txtEmail.Text = cliente.Email ?? string.Empty;
        txtRepresentante.Text = cliente.Representante ?? string.Empty;
        dgvDireccion.Rows.Clear();
        foreach (var direccion in cliente.Direcciones)
            dgvDireccion.Rows.Add(direccion.Calle, direccion.Ciudad, direccion.Provincia);

        SetDetailInputsEnabled(false);
        btnNuevo.Enabled = true;
        btnGrabar.Enabled = false;
        btnEditar.Enabled = true;
        btnCancelar.Enabled = true;
        btnGrabar.Text = "Grabar";
    }

    private void btnNuevo_Click()
    {
        _isCreateMode = true;
        tabPageDetalles.Enabled = true;
        tabPageListado.Enabled = false;
        tabControl1.SelectTab(tabPageDetalles);
        ClearDetails();
        SetDetailInputsEnabled(true);
        txtCodigo.ReadOnly = true;
        txtNombre.Focus();
        btnNuevo.Enabled = false;
        btnGrabar.Enabled = true;
        btnCancelar.Enabled = true;
        btnEditar.Enabled = false;
        btnGrabar.Text = "Grabar";
    }

    private void btnEditar_Click()
    {
        SetDetailInputsEnabled(true);
        txtCodigo.ReadOnly = true;
        txtNombre.Focus();
        btnNuevo.Enabled = false;
        btnGrabar.Enabled = true;
        btnEditar.Enabled = false;
        btnCancelar.Enabled = true;
        btnGrabar.Text = "Actualizar";
        _isCreateMode = false;
    }

    private async Task ReturnToListingAsync()
    {
        tabPageListado.Enabled = true;
        tabPageDetalles.Enabled = false;
        tabControl1.SelectTab(tabPageListado);
        _isCreateMode = true;
        btnGrabar.Text = "Grabar";
        btnNuevo.Enabled = true;
        btnGrabar.Enabled = false;
        btnEditar.Enabled = false;
        btnCancelar.Enabled = false;
        optBuscar.Checked = true;
        _clientes = await Mediator.Send(new ListClientesQuery());
        dgvListado.Rows.Clear();
    }

    private async Task EnsureTiposIdentificacionAsync()
    {
        if (_tiposIdentificacion.Count > 0) return;
        _tiposIdentificacion = await Mediator.Send(new ListTiposIdentificacionQuery());
        cbTipoID.DataSource = _tiposIdentificacion.ToList();
        cbTipoID.DisplayMember = nameof(TipoIdentificacionDto.Descripcion);
        cbTipoID.ValueMember = nameof(TipoIdentificacionDto.Codigo);
    }

    private void SelectTipoIdentificacion(int codigo)
    {
        if (_tiposIdentificacion.Count == 0) return;
        cbTipoID.SelectedValue = codigo;
        if (!Equals(cbTipoID.SelectedValue, codigo))
            cbTipoID.Text = _tiposIdentificacion.FirstOrDefault(tipo => tipo.Codigo == codigo)?.Descripcion ?? string.Empty;
    }

    private TipoIdentificacionDto? ResolveTipoIdentificacion()
    {
        if (cbTipoID.SelectedValue is not null)
        {
            var codigo = Convert.ToInt32(cbTipoID.SelectedValue, CultureInfo.InvariantCulture);
            return _tiposIdentificacion.FirstOrDefault(tipo => tipo.Codigo == codigo);
        }

        return _tiposIdentificacion.FirstOrDefault(tipo =>
            string.Equals(tipo.Descripcion, cbTipoID.Text, StringComparison.CurrentCultureIgnoreCase));
    }

    private void cbTipoID_SelectedIndexChanged()
    {
        txtNroID.Clear();
        if (string.Equals(cbTipoID.Text, "sin documento", StringComparison.CurrentCultureIgnoreCase))
            txtNroID.Text = "00000000";
    }

    private void txtNroID_GotFocus()
    {
        txtNroID.MaxLength = 32767;
        if (string.Equals(cbTipoID.Text, "dni", StringComparison.CurrentCultureIgnoreCase))
            txtNroID.MaxLength = 8;
        if (string.Equals(cbTipoID.Text, "ruc", StringComparison.CurrentCultureIgnoreCase))
            txtNroID.MaxLength = 11;
    }

    private async Task txtBuscarCodigo_KeyPress(KeyPressEventArgs e)
    {
        if (e.KeyChar != (char)Keys.Enter) return;
        e.Handled = true;
        dgvListado.Rows.Clear();
        if (!int.TryParse(txtBuscarCodigo.Text, NumberStyles.Integer, CultureInfo.InvariantCulture, out var codigo)) return;
        var cliente = await Mediator.Send(new GetClienteByIdQuery(codigo));
        if (cliente is not null)
            BindClientes([cliente]);
    }

    private void ClearDetails()
    {
        txtCodigo.Clear();
        txtNombre.Clear();
        cbTipoID.SelectedIndex = -1;
        txtNroID.Clear();
        txtTelefono.Clear();
        txtFax.Clear();
        txtEmail.Clear();
        txtRepresentante.Clear();
        dgvDireccion.Rows.Clear();
    }

    private void SetDetailInputsEnabled(bool enabled)
    {
        txtCodigo.Enabled = enabled;
        txtNombre.Enabled = enabled;
        cbTipoID.Enabled = enabled;
        txtNroID.Enabled = enabled;
        txtTelefono.Enabled = enabled;
        txtFax.Enabled = enabled;
        txtEmail.Enabled = enabled;
        txtRepresentante.Enabled = enabled;
        dgvDireccion.Enabled = enabled;
    }

    private IReadOnlyList<DireccionDto> ReadDirecciones()
    {
        var direcciones = new List<DireccionDto>();
        foreach (DataGridViewRow row in dgvDireccion.Rows)
        {
            if (row.IsNewRow) continue;
            var calle = Convert.ToString(row.Cells[0].Value, CultureInfo.CurrentCulture) ?? string.Empty;
            var ciudad = Convert.ToString(row.Cells[1].Value, CultureInfo.CurrentCulture) ?? string.Empty;
            var provincia = Convert.ToString(row.Cells[2].Value, CultureInfo.CurrentCulture) ?? string.Empty;
            if (string.IsNullOrWhiteSpace(calle) && string.IsNullOrWhiteSpace(ciudad) && string.IsNullOrWhiteSpace(provincia)) continue;
            direcciones.Add(new DireccionDto(calle, ciudad, provincia));
        }
        return direcciones;
    }

    private void BindClientes(IEnumerable<ClienteDto> clientes)
    {
        dgvListado.Rows.Clear();
        foreach (var cliente in clientes)
        {
            var rowIndex = dgvListado.Rows.Add();
            var row = dgvListado.Rows[rowIndex];
            row.Cells[0].Value = cliente.Codigo;
            row.Cells[1].Value = cliente.RazonSocial;
            row.Cells[2].Value = GetTipoIdentificacionDescripcion(cliente.TipoIdCodigo);
            row.Cells[3].Value = cliente.NroIdentificacion;
        }
    }

    private string GetTipoIdentificacionDescripcion(int codigo) =>
        _tiposIdentificacion.FirstOrDefault(tipo => tipo.Codigo == codigo)?.Descripcion ?? string.Empty;

    private bool ShowFailure(string? error, bool showMessages)
    {
        var message = error ?? "No se pudo guardar el cliente.";
        SetStatus(message);
        if (showMessages) MessageBox.Show(message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
        return false;
    }

    private void SetStatus(string message) => lblmsg.Text = message;

    private static string? EmptyToNull(string value) => string.IsNullOrWhiteSpace(value) ? null : value.Trim();

    private IMediator Mediator => _mediator
        ?? throw new InvalidOperationException("Use the IMediator constructor at runtime. The parameterless constructor is for the WinForms designer only.");
}
